--- code/D3MFImporter.cpp	2018-05-13 23:05:48.421253063 +0200
+++ code/D3MFImporter.cpp	2018-05-13 23:03:57.314014344 +0200
@@ -227,7 +227,7 @@
         aiVector3D vertex;
         vertex.x = ai_strtof(xmlReader->getAttributeValue(D3MF::XmlTag::x.c_str()), nullptr);
         vertex.y = ai_strtof(xmlReader->getAttributeValue(D3MF::XmlTag::y.c_str()), nullptr);
-        vertex.z = ai_strtof>(xmlReader->getAttributeValue(D3MF::XmlTag::z.c_str()), nullptr);
+        vertex.z = ai_strtof(xmlReader->getAttributeValue(D3MF::XmlTag::z.c_str()), nullptr);
 
         return vertex;
     }
