diff -u -r -N src.patched/Standard/Standard_CLocaleSentry.hxx src/Standard/Standard_CLocaleSentry.hxx
--- src.patched/Standard/Standard_CLocaleSentry.hxx	2017-08-11 07:51:11.000000000 +0200
+++ src/Standard/Standard_CLocaleSentry.hxx	2017-10-23 10:53:43.571058000 +0200
@@ -29,9 +29,11 @@
     #define HAVE_XLOCALE_H
   #endif
 
-  //! We check _GNU_SOURCE for glibc extensions here and it is always defined by g++ compiler.
+  //! We check _GNU_SOURCE for glibc <= v2.25 extensions here and it is always defined by g++ compiler.
   #if defined(_GNU_SOURCE) && !defined(__ANDROID__)
-    #define HAVE_XLOCALE_H
+    #if defined(__GLIBC__) && __GLIBC__ == 2 && __GLIBC_MINOR__ <= 25
+      #define HAVE_XLOCALE_H
+    #endif
   #endif
 #endif // ifndef HAVE_LOCALE_H
 
